### PR TITLE
simgrid: 3.28 -> 3.29 + python + clean

### PR DESCRIPTION
**Pull request by mpoquet**
This commit makes nixpkgs simgrid in sync with NUR-Kapack's.
- use latest SimGrid release
- enable the python binding (in new output "python")

[Here](https://github.com/oar-team/nur-kapack/commits/57f34c9de41d6ffe71b74ea0b87e3cf52fb81361/pkgs/simgrid/simgrid329.nix) is the detailed history of changes in NUR-Kapack that led to this commit.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.
- [x] Checked that closure size did not increase.

### CODE REVIEW

File: pkgs/applications/science/misc/simgrid/default.nix [hunk 1]
```diff
@@ -1,13 +1,14 @@
-{ lib, stdenv, fetchFromGitLab, fetchpatch, cmake, perl, python3, boost, valgrind
```
[the remark this answers is not in the document]
Comment (mpoquet) [reply pkgs/applications/science/misc/simgrid/default.nix:1]: Yes, valgrind does not seem to be required anymore. There is no valgrind required by SimGrid's build system nor runtime code, and SimGrid main devs told me that it should not be required now.

File: pkgs/applications/science/misc/simgrid/default.nix [hunk 2]
```diff
@@ -18,54 +19,34 @@ in
...
 
-  #buildInputs = optional luaSupport lua5;
-
-  # Make it so that libsimgrid.so will be found when running programs from
-  # the build dir.
-  preConfigure = ''
-    export LD_LIBRARY_PATH="$PWD/build/lib"
-  '';
+  outputs = ["out"]
+    ++ optionals buildPythonBindings ["python"];
```
[the remark this answers is not in the document]
Comment (mpoquet) [reply pkgs/applications/science/misc/simgrid/default.nix:41]: The `out` output is a store dependency of the `python` output.
My motivation to generate two outputs is to reduce the closure size of SimGrid users (e.g., [Batsim](https://framagit.org/batsim/batsim/)) that only want the C++ library and do not care about the python binding.
This setup makes this work without forcing users to recompile SimGrid with `buildPythonBindings` set to false.

File: pkgs/applications/science/misc/simgrid/default.nix [hunk 3]
```diff
@@ -75,27 +56,28 @@ stdenv.mkDerivation rec {
...
+    "-Denable_compile_warnings=off"
+    "-Denable_compile_optimizations=${optionOnOff optimize}"
+    "-Denable_lto=${optionOnOff optimize}"
   ];
-
   makeFlags = optional debug "VERBOSE=1";
 
-  # Some Perl scripts are called to generate test during build which
-  # is before the fixupPhase, so do this manualy here:
+  # needed to ensure correct perl dependency in outputs
```
[the remark this answers is not in the document]
Comment (mpoquet) [reply pkgs/applications/science/misc/simgrid/default.nix]: I updated the comment and made it simpler. Calling `patchShebangs` is no longer needed at build/test time, so I moved the call to the fixupPhase to just fix the generated output scripts.

File: pkgs/applications/science/misc/simgrid/default.nix [hunk 4]
```diff
@@ -75,27 +56,28 @@ stdenv.mkDerivation rec {
...
   preBuild = ''
     patchShebangs ..
```
[the remark this answers is not in the document]
Comment (mpoquet) [reply pkgs/applications/science/misc/simgrid/default.nix]: I only patched the script files in `$out`, as doing the patch before build/test is no longer needed. I therefore moved the call to `patchShebangs` to the fixup phase.

File: pkgs/applications/science/misc/simgrid/default.nix [hunk 5]
```diff
@@ -104,6 +86,24 @@ stdenv.mkDerivation rec {
...
+  postInstall = "" + lib.optionalString withoutBin ''
+    # remove bin from output if requested.
+    # having a specific bin output would be cleaner but it does not work currently (circular references)
+    rm -rf $out/bin
+  '' + lib.optionalString buildPythonBindings ''
+    # manually install the python binding if requested.
+    # library output file is expected to be formatted as simgrid.cpython-XY-.*.so
+    # where X is python version major and Y python version minor.
+    PYTHON_VERSION_MAJOR=$(ls lib/simgrid.cpython*.so | sed -E 'sWlib/simgrid\.cpython-([[:digit:]])([[:digit:]])-.*W\1W')
+    PYTHON_VERSION_MINOR=$(ls lib/simgrid.cpython*.so | sed -E 'sWlib/simgrid\.cpython-([[:digit:]])([[:digit:]])-.*W\2W')
```
[the remark this answers is not in the document]
Comment (mpoquet) [reply pkgs/applications/science/misc/simgrid/default.nix]: Thanks, I did not know that this existed :).

**Comment by mpoquet** [comment]
Thanks for the reviews! I just pushed 02c1160b7056cf5a to apply them.

**Review by mickours — APPROVED** [review]

**Comment by mpoquet** [comment]
Superseded by #157617.

**Pull request closed without merging**